File: array/128.py
#for loop + while loop看似两层循环, 但是while loop只会出现在n-1 is None的情况下 所以 整体时间复杂度O(n)
class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        num_set = set(nums)
        
        ret = 0
        
        for num in num_set:
            
            if num - 1 not in num_set:
                cur_len = 1
            
                while num + 1 in num_set:
                    cur_len += 1
                    num += 1
                
                
                ret = max(ret, cur_len)
            
        return ret
            

File: array/128.cc
class Solution {
public:
    int longestConsecutive(vector<int>& nums) {
        std::unordered_set<int> s;
        for (const auto &num : nums) {
            s.emplace(num);
        }
        int ret = 0;
        for (const auto &num : nums) {
            // 检查这个num 是否已经在一个答案序列里, 如果不在 以这个为起点检查答案序列
            auto iter = s.find(num-1);
            if (iter == s.end()) {
                int cur = num;
                int count = 1;
                while (true) {
                    ++cur;
                    auto iter_cur = s.find(cur);
                    if (iter_cur != s.end()) {
                        ++count;
                    } else {
                        break;
                    }
                }
                ret = std::max(count, ret);
            }
        }     
        return ret;
    }
};
